### PR TITLE
Backend channel numbers api matrix & undelete/trashcan

### PR DESCRIPTION
**Pull request by phunkyfish**
v4.4.0
- Added: Set program id option for streams with superfluous program data
- Added: Undelete and trashcan (when configured on backend) for recordings
- Added: Use new API for backend channel numbers - openwebif 1.3.7
- Fixed: Radio groups parsed from wrong api
- Added: Support disabling addon debug logging in debug mode

### CODE REVIEW

**Comment by phunkyfish** [comment]
I left disabling debug output as an option as I've actually used it to good effect when debugging the new tvOS builds.

**Comment by phunkyfish** [comment]
Assuming that failing apple builds are related to: xbmc/xbmc#16218

**Comment by ksooo** [comment]
> Assuming that failing apple builds are related to: xbmc/xbmc#16218

 Yep. @Rechi any news regarding a fix for that?

**Comment by phunkyfish** [comment]
@ksooo should be ready to go.

**Comment by phunkyfish** [comment]
@ksooo now it’s really ready 😉

File: pvr.vuplus/resources/language/resource.language.en_gb/strings.po [hunk 1]
```diff
@@ -721,7 +721,12 @@ msgctxt "#30143"
 msgid "Radio bouquet 5"
 msgstr ""
 
-#empty strings from id 30144 to 30409
+#label: Advanced - ignoredebug
+msgctxt "#30144"
+msgid "Ignore debug logging in debug mode"
```
Comment (ksooo) [inline pvr.vuplus/resources/language/resource.language.en_gb/strings.po]: "No addon debug logging in Kodi debug mode" ?

File: src/enigma2/ChannelGroups.cpp [hunk 2]
```diff
@@ -100,7 +102,20 @@ std::shared_ptr<ChannelGroup> ChannelGroups::GetChannelGroup(int uniqueId)
   return m_channelGroups.at(uniqueId - 1);
 }
 
-std::shared_ptr<ChannelGroup> ChannelGroups::GetChannelGroup(std::string groupName)
+std::shared_ptr<ChannelGroup> ChannelGroups::GetChannelGroup(const std::string &groupServiceReference)
+{
+  std::shared_ptr<ChannelGroup> channelGroup = nullptr;
```
Comment (ksooo) [inline src/enigma2/ChannelGroups.cpp]: No need for `= nullptr` here.

File: src/enigma2/ChannelGroups.cpp [hunk 3]
```diff
@@ -100,7 +102,20 @@ std::shared_ptr<ChannelGroup> ChannelGroups::GetChannelGroup(int uniqueId)
...
+
+  auto channelGroupPair = m_channelGroupsServiceReferenceMap.find(groupServiceReference);
```
Comment (ksooo) [inline src/enigma2/ChannelGroups.cpp]: Ha, my favourite again: `const auto`.

File: src/enigma2/ChannelGroups.cpp [hunk 4]
```diff
@@ -100,7 +102,20 @@ std::shared_ptr<ChannelGroup> ChannelGroups::GetChannelGroup(int uniqueId)
...
+  if (channelGroupPair != m_channelGroupsServiceReferenceMap.end())
+  {
+    channelGroup = channelGroupPair->second;
+  }
+
+  return channelGroup;
+}
```
Comment (ksooo) [inline src/enigma2/ChannelGroups.cpp:112]: I would write this a bit different (shorter, and functionally identical):

```c++
std::shared_ptr<ChannelGroup> ChannelGroups::GetChannelGroup(const std::string &groupServiceReference)
{
  const auto channelGroupPair = m_channelGroupsServiceReferenceMap.find(groupServiceReference);
  if (channelGroupPair != m_channelGroupsServiceReferenceMap.end())
    return channelGroupPair->second;

  return {};
}
```

File: src/enigma2/Recordings.cpp [hunk 5]
```diff
@@ -412,7 +427,7 @@ void Recordings::SetRecordingNextSyncTime(RecordingEntry &recordingEntry, time_t
 
 PVR_ERROR Recordings::DeleteRecording(const PVR_RECORDING &recinfo)
 {
-  const std::string strTmp = StringUtils::Format("web/moviedelete?sRef=%s", WebUtils::URLEncodeInline(recinfo.strRecordingId).c_str());
```
Comment (ksooo) [inline src/enigma2/Recordings.cpp]: why did you remove `const`here?

File: src/enigma2/Recordings.cpp [hunk 6]
```diff
@@ -424,6 +439,37 @@ PVR_ERROR Recordings::DeleteRecording(const PVR_RECORDING &recinfo)
...
+PVR_ERROR Recordings::UndeleteRecording(const PVR_RECORDING& recording)
+{
+  auto recordingEntry = GetRecording(recording.strRecordingId);
+
+  std::regex regex(TRASH_FOLDER);
+  std::string replaceWith = "";
+
+  const std::string newRecordingDirectory = regex_replace(recordingEntry.GetDirectory(), regex, replaceWith);
+
+  std::string strTmp = StringUtils::Format("web/moviemove?sRef=%s&dirname=%s", WebUtils::URLEncodeInline(recordingEntry.GetRecordingId()).c_str(), WebUtils::URLEncodeInline(newRecordingDirectory).c_str());
```
Comment (ksooo) [inline src/enigma2/Recordings.cpp]: const std::string?

File: src/enigma2/Recordings.cpp [hunk 7]
```diff
@@ -424,6 +439,37 @@ PVR_ERROR Recordings::DeleteRecording(const PVR_RECORDING &recinfo)
...
 }
 
+PVR_ERROR Recordings::UndeleteRecording(const PVR_RECORDING& recording)
+{
+  auto recordingEntry = GetRecording(recording.strRecordingId);
+
+  std::regex regex(TRASH_FOLDER);
+  std::string replaceWith = "";
+
+  const std::string newRecordingDirectory = regex_replace(recordingEntry.GetDirectory(), regex, replaceWith);
```
Comment (ksooo) [inline src/enigma2/Recordings.cpp]: Do you really need var `repaceWith`? Can't you just call `regex_replace` with "" as last parameter?

File: src/enigma2/Recordings.h [hunk 8]
```diff
@@ -38,42 +38,46 @@ namespace enigma2
   static int CUTS_LAST_PLAYED_TYPE = 3;
   static int E2_DEVICE_LAST_PLAYED_SYNC_INTERVAL_MIN = 300;
   static int E2_DEVICE_LAST_PLAYED_SYNC_INTERVAL_MAX = 600;
+  static std::string TRASH_FOLDER = ".Trash";
```
Comment (ksooo) [inline src/enigma2/Recordings.h]: constexpr static std::string

File: src/enigma2/data/RecordingEntry.h [hunk 9]
```diff
@@ -106,6 +109,7 @@ namespace enigma2
       mutable bool m_radio = false;
       mutable bool m_haveChannelType = false;
       mutable bool m_anyChannelTimerSource = false;
+      mutable bool m_deleted = false;
```
Comment (ksooo) [inline src/enigma2/data/RecordingEntry.h]: As far as I can see there is no need to flag `m_deleted` mutable as you do not modify it in any `const` member function.

File: pvr.vuplus/resources/language/resource.language.en_gb/strings.po (hunk 1 above)
Comment (phunkyfish) [reply pvr.vuplus/resources/language/resource.language.en_gb/strings.po]: It's an option to switch it off. Used it quite a bit when testing TVOS (where you're live debugging in xcode).

Comment (ksooo) [reply pvr.vuplus/resources/language/resource.language.en_gb/strings.po]: Sure, I understand the semantics. My comment is about wording. I felt that inserting "Kodi" makes it a bit more clear what this setting does.

File: src/enigma2/ChannelGroups.cpp (hunk 3 above)
Comment (phunkyfish) [reply src/enigma2/ChannelGroups.cpp]: Ahhhhhh, one day I'll start getting this right :(

File: src/enigma2/ChannelGroups.cpp (hunk 4 above)
Comment (phunkyfish) [reply src/enigma2/ChannelGroups.cpp:112]: So with this `return {};` it will essentially return `nullptr`?

Comment (ksooo) [reply src/enigma2/ChannelGroups.cpp:112]: Yep. It's the same as `return std::shared_ptr<ChannelGroup>();`

File: src/enigma2/Recordings.cpp (hunk 7 above)
Comment (phunkyfish) [reply src/enigma2/Recordings.cpp]: Nope

File: src/enigma2/ChannelGroups.cpp (hunk 4 above)
Comment (phunkyfish) [reply src/enigma2/ChannelGroups.cpp:112]: Cool

File: pvr.vuplus/resources/language/resource.language.en_gb/strings.po (hunk 1 above)
Comment (phunkyfish) [reply pvr.vuplus/resources/language/resource.language.en_gb/strings.po]: Gotcha, makes sense

File: src/enigma2/Recordings.h (hunk 8 above)
Comment (phunkyfish) [reply src/enigma2/Recordings.h]: This would have an error: `constexpr variable cannot have non-literal type 'const std::string'`

Comment (ksooo) [reply src/enigma2/Recordings.h]: What speaks against making it char*?

Comment (phunkyfish) [reply src/enigma2/Recordings.h]: You mean still using constexpr? Then you get this: `warning: ISO C++11 does not allow conversion from string literal to 'char *const' [-Wwritable-strings]`

Comment (ksooo) [reply src/enigma2/Recordings.h]: `static constexpr const char* TRASH_FOLDER = ".Trash";`

Comment (phunkyfish) [reply src/enigma2/Recordings.h]: This works: `constexpr static const char*`!

**Comment by phunkyfish** [comment]
@ksooo ok, good to go!

File: src/enigma2/Recordings.h (hunk 8 above)
Comment (ksooo) [reply src/enigma2/Recordings.h]: Starting with 'static' is more common, but this is only sugar ofc. => static constexpr const char*

**Comment by phunkyfish** [comment]
Ok, good to go (with sugar) ;)

**Review by ksooo — APPROVED** [review]
👍
